cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/ForwardingDatasetFramework.java

```diff
@@ -0,0 +1,199 @@
+/*
+ * Copyright © 2016 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package co.cask.cdap.internal.app.runtime.workflow;
+
+import co.cask.cdap.api.dataset.Dataset;
+import co.cask.cdap.api.dataset.DatasetAdmin;
+import co.cask.cdap.api.dataset.DatasetProperties;
+import co.cask.cdap.api.dataset.DatasetSpecification;
+import co.cask.cdap.api.dataset.module.DatasetModule;
+import co.cask.cdap.api.workflow.Workflow;
+import co.cask.cdap.data2.datafabric.dataset.type.DatasetClassLoaderProvider;
+import co.cask.cdap.data2.dataset2.DatasetFramework;
+import co.cask.cdap.data2.dataset2.DatasetManagementException;
+import co.cask.cdap.data2.metadata.writer.ProgramContextAware;
+import co.cask.cdap.proto.DatasetSpecificationSummary;
+import co.cask.cdap.proto.Id;
+import org.apache.twill.filesystem.Location;
+
+import java.io.IOException;
+import java.util.Collection;
+import java.util.Map;
+import javax.annotation.Nullable;
+
+/**
+ * Implementation of the {@link DatasetFramework} which forwards all calls to the underlying
+ * dataset framework. Before forwarding the calls, local dataset names are mapped to the names
+ * assigned to them for the current run of the {@link Workflow}.
+ */
+public class ForwardingDatasetFramework implements DatasetFramework, ProgramContextAware {
```

Comment:
By convention of forwarding object, the `ForwardingDatasetFramework` shouldn't implements `ProgramContextAware` and shouldn't have logic for dataset name mapping. The class hierarchy should be like this:

  * `FowardingDatasetFramework implments DatasetFramework` - Simply delegate all methods to another instance
  * `WorkflowDatasetFramework extends ForwardingDatasetFramework implements ProgramContextAware` - Do the local dataset name mapping by overriding methods. It also implements the `ProgramContextAware` methods and do forwarding in case the delegate is an instance of `ProgramContextAware`.
  * The existing `LineageWriterDatasetFramework` should extends from `ForwardingDatasetFramework` as well to reduce code duplication.